File: src/logic/CrosswordFiller.java
package logic;

import model.Crossword;

import java.util.ArrayList;
import java.util.List;

/**
 * User: Dmytro Vynokurov
 * Date: 9/29/13
 * Time: 8:20 PM
 */
public class CrosswordFiller {
    private List<Crossword> variants;
    private int numberOfWords;

    public CrosswordFiller(Crossword crossword) {
        this.variants = new ArrayList<>();
        this.variants.add(crossword);
        this.numberOfWords = crossword.getWords().size();
    }

    public List<Crossword> fillCrossword(){
        for(int i=0;i<numberOfWords;i++){
            matchAllWordsForAllCrosswords();
        }

        return variants;
    }

    private void matchAllWordsForAllCrosswords(){
        OneWordFiller oneWordFiller = new OneWordFiller();
        List<Crossword> newVariants = new ArrayList<>();

        for(Crossword crossword: variants){
            oneWordFiller.matchAllWordsForCrossword(crossword,newVariants);
        }

        for(Crossword c:variants){
            System.out.println(c);
        }
        System.out.println("**********************************************");


        variants = newVariants;
    }


}

File: src/functional/logic/CrosswordFiller.scala
package functional.logic

import model.Crossword
import OneWordFiller._

/**
 * User: Dmytro Vynokurov
 * Date: 12/9/13
 * Time: 12:15 AM
 */
object CrosswordFiller {

  def fillCrossword(crossword:Crossword):List[Crossword] = {
    nextGeneration(crossword::Nil,crossword.getWords.size)
  }

  private def nextGeneration(variants: List[Crossword],generationsLeft:Int):List[Crossword] = {
    if(generationsLeft==0) variants
    else {
      val newVariants = matchAllWordsForAllVariants(variants)
      for(variant<-variants)println(variant)
      println("*************************************")
      nextGeneration(newVariants,generationsLeft-1)
    }
  }

  private def matchAllWordsForAllVariants(variants:List[Crossword]):List[Crossword] = variants match {
    case x::xs => matchAllWordsForCrossword(x):::matchAllWordsForAllVariants(xs)
    case Nil => Nil
  }

}
